kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaStreamingSourceStateStoreRecoveryTest.java

```diff
@@ -0,0 +1,239 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.kafka;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.gson.Gson;
+import io.cdap.cdap.api.app.AppStateStore;
+import io.cdap.cdap.api.artifact.ArtifactRange;
+import io.cdap.cdap.api.artifact.ArtifactSummary;
+import io.cdap.cdap.api.artifact.ArtifactVersion;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.dataset.table.Table;
+import io.cdap.cdap.common.utils.Networks;
+import io.cdap.cdap.common.utils.Tasks;
+import io.cdap.cdap.datastreams.DataStreamsApp;
+import io.cdap.cdap.datastreams.DataStreamsSparkLauncher;
+import io.cdap.cdap.etl.api.streaming.StreamingSource;
+import io.cdap.cdap.etl.mock.batch.MockSink;
+import io.cdap.cdap.etl.mock.test.HydratorTestBase;
+import io.cdap.cdap.etl.mock.transform.RecoveringTransform;
+import io.cdap.cdap.etl.proto.v2.DataStreamsConfig;
+import io.cdap.cdap.etl.proto.v2.ETLPlugin;
+import io.cdap.cdap.etl.proto.v2.ETLStage;
+import io.cdap.cdap.proto.ProgramRunStatus;
+import io.cdap.cdap.proto.artifact.AppRequest;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.ArtifactId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.test.ApplicationManager;
+import io.cdap.cdap.test.DataSetManager;
+import io.cdap.cdap.test.SparkManager;
+import io.cdap.cdap.test.TestBase;
+import io.cdap.cdap.test.TestConfiguration;
+import io.cdap.plugin.common.http.HTTPPollConfig;
+import io.cdap.plugin.kafka.batch.source.KafkaPartitionOffsets;
+import io.cdap.plugin.kafka.source.KafkaStreamingSource;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.spark.streaming.kafka010.KafkaUtils;
+import org.apache.zookeeper.server.ServerCnxnFactory;
+import org.apache.zookeeper.server.ZooKeeperServer;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Tests for Kafka streaming source with state store
+ */
+public class KafkaStreamingSourceStateStoreRecoveryTest extends HydratorTestBase {
```

Comment:
is there a reason we need two classes instead of putting them both in the same class? There's quite a bit of overhead associated with each test class, as they need to startup all the CDAP services.

Reply:
This is because of the classloading issue we saw were Kafka was having static references to org.apache.kafka.common.serialization.ByteArrayDeserializer class. (https://cdap.atlassian.net/browse/PLUGIN-1414)

Reply:
ah I see. Can you add a comment about that?

Reply:
Done